### DISCUSSION TITLE
Running scheduled tasks alongside Rocket

### DISCUSSION THREAD

**Discussion by djc-id** [rwf2/Rocket#1985] [General]
A possibly interesting Reddit thread 

https://www.reddit.com/r/rust/comments/qsgxs2/scheduled_background_task_using_database_in_rocket/

I love Rocket but it does need to coexist in a wider ecosystem and there may be an idiom here that is worth documenting. 

I don't know enough about Rocket or Tokio to confidently document the optimum solution. Let me know if I can help though.

**Comment by tobymurray** [comment]
It looks like this is not resolved, `on_liftoff` needs to return for Rocket to continue so going into an "infinite loop" (a background task that's expected to run for the duration of the application) won't work invoked in the way the post mentions.

**Reply by djc-id** [reply]
The on_liftoff approach doesn't work but the tokio-cron-scheduler approach mentioned later in that thread does.

**Reply by tobymurray** [reply]
Using `tokio-cron-scheduler` is not really Rocket related, which is maybe your original point. It only uses the configuration of Rocket then it's a separate database connection and a separate thread that just so happens to be running near a Rocket instance. Especially if it were to need access to e.g. managed state, a background job like that would get awfully messy pretty quick.

**Comment by somehowchris** [comment]
@djc-id why would you include basically cron jobs or tasks in an api? It just doesn‘t scale….


It‘s not about rust, rocket or tokio. Let‘s begin with why would you engineer such a thing? Would you eat a blue banana?

- Cronjobs are part of your infrastructure, cf-task-jobs, kWs cronjobs, crontab….
- handling tasks isn‘t the job of an api, if you need to offload jobs, get redis pub/sub, rabbitmq or kafka

**Reply by tobymurray** [reply]
Not who you responded to, but I'm interested in the problem. I'm not understanding your comment - would you mind elaborating?

> why would you include basically cron jobs or tasks in an api?

Are you questioning whether some type of background jobs/work that is executed out of band from the request-response flow is suitable in Rocket itself? Or are you referring to applications built with Rocket? My impression is that scheduled jobs as part of a web server is a pretty standard bit of functionality (e.g. do some data clean up at off-peak hours or do some data aggregation at the end of business day or similar), so I feel like I'm missing your point.

> Would you eat a blue banana?

They're apparently [extremely delicious](https://duckduckgo.com/?q=blue+java+banana+tree&t=ffab&iar=images&iax=images&ia=images) so, yeah.

**Reply by somehowchris** [reply]
Ok my bad 😅 lovely banana. 

Let me put it that way, giving a snake wheels wouldn’t make it a car, right?

yes, sometimes you need to have background/cron jobs. For example to send mails, cleanup or aggregate data. Never doubted that, never will. What I’m saying is that you shouldn’t build the job execution into your api Server. The functionality, yes but not the part actually executing it. Let that be the job of cf, k8s, crontab or what ever. But why? Imagine you scale your app, suddenly you send these mails twice

**Reply by MCOfficer** [reply]
I've written up a longer answer, but come to realize it's as simple as "I don't expect to scale beyond one instance".

There's also some things like cache refreshes that need to be done for each instance, and which I don't want to block requests for. Sure, I *can* wrap my cache in an `Arc`, use that as state and have a separate thread that updates it, but it's not exactly pretty.

**Reply by somehowchris** [reply]
@MCOfficer So you are risking to flush your cache during a request? rocket usually has `# of thread(s) * 2` workers.

Sure if you do not scale beyond one instance, but let me tell you from MY (sry, just wanna make sure its clear it mine and not everyone agrees ) experience. First of all its just a bad habbit for professional systems, second, its hell of a lot more complicated if you go on and expand you system with an "all in one" solution.

**Reply by MCOfficer** [reply]
"Cache" was probably a bit ambiguous. Think some form of data in an `RwLock` that gets refreshed every now and then, not as consequence of a request but from somewhere else.

As for professional systems: can't agree more, I would 100% advocate for this while on the clock. But when I'm programming for funzies, I really don't expect any form of scaling.

**Reply by tobymurray** [reply]
> What I’m saying is that you shouldn’t build the job execution into your api Server.

I think I'm in strong agreement with @MCOfficer - in my experience the vast majority of projects would be thrilled to get to the point where they're deployed at all, a slice of those would love to get more than the author as a user, a yet smaller portion get used by a few extremely targeted people, and vanishingly few overall actually reach the point they need to consider scaling.

Personally, I use a single 5$/month VPS to host all my (not at all impressive) work and the only issue has been disk space so far, which could be resolved for the next few years for another ~2$/month. Introducing complexity to address scaling concerns is YAGNI/premature optimization territory in my opinion.

The existence of e.g. [Active Job](https://guides.rubyonrails.org/active_job_basics.html), [Laravel task scheduling](https://laravel.com/docs/8.x/scheduling), [Quartz](http://www.quartz-scheduler.org/), [Django jobs](https://django-extensions.readthedocs.io/en/latest/jobs_scheduling.html) suggest this concept is not completely decoupled from reality.

> Let that be the job of cf, k8s, crontab or what ever.

Using crontab for something like this is not easily maintainable. Doing host-specific non-source-controlled changes is a recipe for trouble, in my opinion, and bucks the trend of infrastructure-as-code. I've never used Cloud Foundry, maybe that's an avenue to pursue, but (following the point above) K8s is overkill for everything I've ever done on my personal time, so it's just not a good fit.

**Reply by somehowchris** [reply]
Yes I get why would you like to do it, am I for it, no definitly not. Would it be more comfortable for you, sure. But comfortability is where mistakes begin.

Let me get onto what you wrote:
 - You mentioned all job scheduling libs/parts for interpreted or IL languages, Rust is not interpreted, not transpiled, has no runtime (besides llvm, which is included)
 - cargo lets you easily have 2 binaries in a repository, let's say you have the server and a small bin just hitting an endpoint
 - well k8s is a thing of its own, k3s can be setup within minutes with no big things, if you need cf you either pay for hosting, you need k8s or a bosh supported vm. So what gives? for crontab it's easy as adding `5 4 * * * docker run xyz:0.0.1` or `5 4 * * * /bin/usr/local/xyz` to the crontab file and restarting the crontab service. `kubectl create cronjob NAME --image=xyz:0.0.1--schedule='5 4 * * * '` for k8s/kubectl. Much more worse for cloud foundry.
 - If you feel like ops is a pain, get used to it, do it over and over again, you learn from what you do _wrong_
 - IaC isn't even code, IaC has to do with weak management and versioning of configurations

**Reply by MCOfficer** [reply]
A few things that I can't help but reply to.

> You mentioned all job scheduling libs/parts for interpreted or IL languages, Rust is not interpreted, not transpiled, has no runtime (besides llvm, which is included)

I feel like I'm missing something here. How is the type of language relevant to how you organize your code?

> cargo lets you easily have 2 binaries in a repository, let's say you have the server and a small bin just hitting an endpoint

If that second binary is making a database connection itself, the way OP described, that makes rocket's DB support useless since the user has to manually manage a pool anyways. Or launch a full rocket instance for that second binary. 

If it just pings an endpoint on the first instance: yikes! Now that's a hack I didn't want to see today. And it's basically the same as this feature request, it still runs an update on rocket's own threadpool, with rocket's managed state. It's just a more convoluted trigger.

> If you feel like ops is a pain, get used to it, do it over and over again, you learn from what you do wrong

If the goal is to become a good overall developer, you're correct. But the goal right here is convenience for the developer. That's why rocket exists in the first place, surely I would learn more by implementing my own server on top of hyper.

**Reply by somehowchris** [reply]
- Ok so why does the language matter, all these mentioned languages have a runtime, a bottleneck for speed, ressources and agility. Rust doesn‘t have one. A rust binary start (basically) instantly, has no runtime, no heap, nothing to manage per se in general.
- Well if you go for real engineering, the second binary wouldn‘t be allowed to access a db or any ressource directly. It would have to connect to the api server. Now it's important you have a idempotent api for these actions.
- It's not about what your jobs do, its about that you as a dev shouldn't implement when your jobs should run, give that to some OPs component, they are responsible for these kind of things. I don't wanna go further but clocks to trigger jobs inside user-land apps aren't terribly off but they aren't really great.

**Reply by MCOfficer** [reply]
Ah, I see your point. Still, your solution being a possibility doesn't entirely invalidate the comparison to other Frameworks. 

I feel like we've come full circle now. Once again, I believe Rocket should be for hobbyists and prototyping just as much as for professional systems. For the first two cases, your solution is pretty much overengineered, which lines up with my own experience - at least 9/10 of my own hobby projects never make it to a point where scaling is even remotely close.

So the actual question is: How much convenience is too much? Given that other, more business-focused frameworks (f.e. Spring) have also implemented this, I feel like this is fine. It's not like this is some hidden antipattern, either - It's fairly obvious it'll go wrong once you want to scale, and it doesn't complicate switching to a more scalable solution. After all, you're only changing the trigger.

**Reply by somehowchris** [reply]
Ok so my last few comments may have been a bit harsh on that. As I said in like the my second comment. I see your point. I see that you don't wanna have a super duper scalable system for 100 million+ users. Well then we would talk about totally different things.

But the last few months I have seen code and system which was and were created just from side projects, devs just copied shit over, never though about principal engineering patterns. Since then I have gone another route. If it's not needed by design, don't do it. No matter what it is, a hobby project, work for some company or institution, doesn't matter. If you create the option to do something wrong, i promise you, it will go wrong. (many devs have also no clue how to test properly which makes things go wrong even worse).

TLDR; do as you like, I don't support it per-se, make sure if you document it to mention engineering patterns

**Reply by MCOfficer** [reply]
That's fair, we can totally agree on "whatever works for you". Offtopic, I feel for you for taking over legacy projects, and I hope you have a nice day :)
